### PR TITLE
Feature/user answer result page integration

### PR DESCRIPTION
**Pull request by elyric-sun-asterisk-dev**
[Commands]
cd backend
php artisan serve

cd frontend
npm start

[Pre-Conditions]
None

[Expected Output]
Expected to take the quiz along with the quiz result failed or passed.

[Notes]
None

[Screenshots]
image
image
image
image

### CODE REVIEW

File: frontend/src/pages/user/UserAnswerPage.js [hunk 1]
```diff
@@ -39,9 +41,36 @@ export default function UserAnswerPage() {
...
+
+  const calculateTotalScore = () => {
+    return (score / totalQuestion).toFixed(2) * 100;
+  };
+
+  useEffect(() => {
+    const totalScore = calculateTotalScore();
+
+    setFailed(totalScore < passingScore ? true : false);
+    setPassed(totalScore >= passingScore ? true : false);
```
Comment (chevyB) [inline frontend/src/pages/user/UserAnswerPage.js]: You can just use 1 variable instead of 2.

File: frontend/src/pages/user/UserAnswerPage.js [hunk 2]
```diff
@@ -39,9 +41,36 @@ export default function UserAnswerPage() {
...
+  const totalQuestion =
+    questions && Array.isArray(questions) ? questions.length : 0;
+  const passingScore = 75;
+
+  const calculateTotalScore = () => {
+    return (score / totalQuestion).toFixed(2) * 100;
+  };
+
+  useEffect(() => {
+    const totalScore = calculateTotalScore();
```
Comment (chevyB) [inline frontend/src/pages/user/UserAnswerPage.js]: All of these calculation must be calculated in BE. For security purposes.

File: frontend/src/pages/user/UserAnswerPage.js (hunk 1 above)
Comment (elyric-sun-asterisk-dev) [reply frontend/src/pages/user/UserAnswerPage.js]: Actually I already tried that sir I get conflicts and bug to my UserAnswerResultPage. And I have no choice but to separate both of them. :bow:

Comment (chevyB) [reply frontend/src/pages/user/UserAnswerPage.js]: @elyric-sun-asterisk-dev i dont get why it has conflict or bug. You have a choice, to fix the bug if it has. So still create 1 variable for this one.

File: frontend/src/pages/user/UserAnswerPage.js [hunk 3]
```diff
@@ -39,9 +42,46 @@ export default function UserAnswerPage() {
...
     questions && Array.isArray(questions) ? questions.length + 1 : 0;
 
+  const totalQuestion =
+    questions && Array.isArray(questions) ? questions.length : 0;
+  const passingScore = 75;
+
+  useEffect(() => {
+    setFailed(scorePercentage < passingScore ? true : false);
+    setPassed(scorePercentage >= passingScore ? true : false);
+  }, [scorePercentage]);
```
Comment (chevyB) [inline frontend/src/pages/user/UserAnswerPage.js]: Remove this useEffect, it is taking a memory. Just add the logic if fail or pass after api response.
Note that only use 1 variable. setIsPassed with boolean value

File: frontend/src/pages/user/UserAnswerPage.js [hunk 4]
```diff
@@ -39,9 +42,46 @@ export default function UserAnswerPage() {
...
+
+  useEffect(() => {
+    const data = {
+      score,
+      totalQuestion,
+    };
+    const calculate = async (data) => {
+      try {
+        const result = await calculateScore(data);
```
Comment (chevyB) [inline frontend/src/pages/user/UserAnswerPage.js]: dont create a variable if you are only using it once
```suggestion
    const calculate = async (data) => {
      try {
        const result = await calculateScore({
      score,
      totalQuestion,
    });
```

File: frontend/src/pages/user/UserAnswerPage.js [hunk 5]
```diff
@@ -39,9 +42,46 @@ export default function UserAnswerPage() {
...
+        console.log(error);
+      }
+    };
+    calculate(data);
+  }, [isDone]);
+
+  const handleAnswer = (question, answer, correctAnswer) => {
+    setCurrentPage(currentPage + 1);
+    setAnswers(answers.concat([answer]));
+    setIsCorrect(isCorrect.concat([answer === correctAnswer ? true : false]));
```
Comment (chevyB) [inline frontend/src/pages/user/UserAnswerPage.js]: You still calculating in frontend if its correct or not. Again, just store all answers and let BE do the checking and calculation.

File: backend/app/Http/Controllers/ScoreController.php [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Illuminate\Http\Request;
+
+class ScoreController extends Controller
+{
+    public function calculateScore(Request $request)
```
Comment (chevyB) [inline backend/app/Http/Controllers/ScoreController.php:9]: I am expecting to save the answers in the DB, why there was no saving here?

File: frontend/src/pages/user/UserAnswerPage.js [hunk 7]
```diff
@@ -39,13 +41,48 @@ export default function UserAnswerPage() {
...
+        });
+        setScorePercentage(result.result);
+        setScore(result.score);
+        setIsCorrect(result.isCorrect);
+      } catch (error) {
+        console.log(error);
+      }
+    };
+    calculate();
+  }, [isDone]);
```
Comment (chevyB) [inline frontend/src/pages/user/UserAnswerPage.js]: 1. You can set right away if passed or not, dont create useEffect and another variable. its taking a memory
2. rename it to response. Its not good to use result.result.
3. Again, much better to calculate the `passed` in BE so pls remove this calculation `response.result >= passingScore ? true : false`in FE. You are doing the calculation in BE already why not do it all there?
4. What is the `isCorrect` variable used for?Seems like its referring to a single question. Maybe rename it like `result`


```suggestion

  useEffect(() => {
    const calculate = async () => {
      try {
        const response = await calculateScore({
          score,
          totalQuestion,
          answers,
          correctAnswers,
        });
        setScorePercentage(response.result);
        setPassed(response.result >= passingScore ? true : false);
        setScore(response.score);
        setIsCorrect(response.isCorrect);
      } catch (error) {
        console.log(error);
      }
    };
    calculate();
  }, [isDone]);
```

File: frontend/src/pages/user/UserAnswerPage.js [hunk 8]
```diff
@@ -39,13 +41,48 @@ export default function UserAnswerPage() {
...
+
+  const handleAnswer = (answer, correctAnswer) => {
+    setCurrentPage(currentPage + 1);
+    setAnswers(answers.concat([answer]));
+    setCorrectAnswers(correctAnswers.concat([correctAnswer]));
```
Comment (chevyB) [inline frontend/src/pages/user/UserAnswerPage.js:95]: Dont build the correct answers on FE. You can just query it on BE for security purposes. You dont want users to modify the data

File: frontend/src/pages/user/UserAnswerPage.js [hunk 9]
```diff
@@ -39,13 +41,48 @@ export default function UserAnswerPage() {
...
+  }, [scorePercentage]);
+
+  useEffect(() => {
+    const calculate = async () => {
+      try {
+        const result = await calculateScore({
+          score,
+          totalQuestion,
+          answers,
+          correctAnswers,
```
Comment (chevyB) [inline frontend/src/pages/user/UserAnswerPage.js]: 1. Why you are passing score when it is just `0`? remove it.
2. Remove the correctAnswers and TotalQuestions, just fetch it from BE.
3. You can just pass the question id to fetch those.

**Review by chevyB — CHANGES_REQUESTED** [review]

File: frontend/src/pages/user/UserAnswerPage.js [hunk 10]
```diff
@@ -93,18 +138,141 @@ export default function UserAnswerPage() {
     );
   });
 
-  const renderResult = () => {
-    return <div>Result Page</div>;
+  const renderPassedPopupResult = () => {
```
Comment (chevyB) [inline frontend/src/pages/user/UserAnswerPage.js]: Create 1 dynamic popup

File: frontend/src/pages/user/UserAnswerPage.js [hunk 11]
```diff
@@ -62,25 +99,33 @@ export default function UserAnswerPage() {
...
+                  }
                   className={buttonClass}
                 >
                   {choice.choiceC}
                 </button>
                 <button
-                  onClick={() => handleAnswer(question, choice.choiceD)}
+                  onClick={() =>
+                    handleAnswer(choice.choiceD, choice.answer.answer)
+                  }
```
Comment (chevyB) [inline frontend/src/pages/user/UserAnswerPage.js]: Just loop the choices. It is much cleaner.

File: frontend/src/pages/user/UserAnswerPage.js (hunk 7 above)
Comment (elyric-sun-asterisk-dev) [reply frontend/src/pages/user/UserAnswerPage.js]: isCorrect is I use to pass on AnswerResultPage to render the status if isCorrect then show correct image, else show wrong image.

File: frontend/src/pages/user/UserAnswerPage.js [hunk 12]
```diff
@@ -39,13 +51,62 @@ export default function UserAnswerPage() {
...
+          answers,
+          correctAnswers,
+          answersQuestion,
+        });
+        setScorePercentage(response.result);
+        setScore(response.score);
+        setIsCorrect(response.isCorrect);
+        setCorrectAnswersOnly(response.correctAnswersOnly);
+        setCorrectAnswerQuestionsOnly(response.correctAnswerQuestionsOnly);
+        setPassed(response.passed);
```
Comment (chevyB) [inline frontend/src/pages/user/UserAnswerPage.js:83]: can we just use 1 variable for all of this, there are so many setters

**Pull request merged**